ziti/cmd/ziti/cmd/edge_controller/login.go

```diff
@@ -80,70 +81,199 @@ func newLoginCmd(f cmdutil.Factory, out io.Writer, errOut io.Writer) *cobra.Comm
-	_, _ = container.SetP(o.Username, "username")
-	_, _ = container.SetP(o.Password, "password")
 
-	body := container.String()
+	absCertPath, err := filepath.Abs(o.Cert)
+	if err == nil {
+		o.Cert = absCertPath
+	}
 
-	jsonParsed, err := util.EdgeControllerLogin(host, o.Cert, body, o.Out, o.OutputJSONResponse, o.edgeOptions.Timeout, o.edgeOptions.Verbose)
+	loginIdentity := &util.RestClientIdentity{
+		Url:       host,
+		Token:     o.Token,
+		LoginTime: time.Now().Format(time.RFC3339),
+		Cert:      o.Cert,
+		ReadOnly:  o.ReadOnly,
+	}
 
+	config, configFile, err := util.LoadRestClientConfig()
 	if err != nil {
 		return err
 	}
 
-	if !jsonParsed.ExistsP("data.token") {
-		return fmt.Errorf("no session token returned from login request to %v. Received: %v", host, jsonParsed.String())
+	id := config.GetIdentity()
+	o.Printf("Saving identity '%v' to %v\n", id, configFile)
+	config.Identities[id] = loginIdentity
+
+	err = util.PersistRestClientConfig(config)
+
+	return err
+}
+
+func (o *loginOptions) ConfigureCerts(host string, ctrlUrl *url.URL) error {
+	isServerTrusted, err := util.IsServerTrusted(host)
+	if err != nil {
+		return err
 	}
 
-	token, ok := jsonParsed.Path("data.token").Data().(string)
+	if isServerTrusted && o.Cert == "" {
+		wellKnownCerts, certs, err := util.GetWellKnownCerts(host)
+		if err != nil {
+			return errors.Wrapf(err, "unable to retrieve server certificate authority from %v", host)
+		}
+
+		savedCerts, certFile, err := util.ReadCert(ctrlUrl.Hostname())
+		if err != nil {
+			return err
+		}
 
-	if !ok {
-		return fmt.Errorf("session token returned from login request to %v is not in the expected format. Received: %v", host, jsonParsed.String())
+		if savedCerts != nil {
+			o.Cert = certFile
+			if !util.AreCertsSame(o, wellKnownCerts, savedCerts) {
+				o.Printf("WARNING: server supplied certificate authority doesn't match cached certs at %v\n", certFile)
+				o.Printf("Issuer: %v\n", certs[0].Issuer.String())
+				replace, err := o.askYesNo("Replace cached certs [Y/N]: ")
+				if err != nil {
+					return err
+				}
+				if replace {
+					_, err = util.WriteCert(o, ctrlUrl.Hostname(), wellKnownCerts)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		} else {
+			if len(certs) == 0 {
+				o.Print("WARNING: server supplied certificate authority was empty, continuing but login will likely fail")
+			} else {
+				o.Printf("Unknown certificate authority retrieved from server\n")
+				o.Printf("Issuer: %v\n", certs[0].Issuer.String())
+				importCerts, err := o.askYesNo("Use server provided certificate authority [Y/N]: ")
+				if err != nil {
+					return err
+				}
+				if importCerts {
+					o.Cert, err = util.WriteCert(o, ctrlUrl.Hostname(), wellKnownCerts)
+					if err != nil {
+						return err
+					}
+				} else {
+					o.Print("WARNING: no certificate authority provided for server, continuing but login will likely fail")
+				}
+			}
+		}
+	} else if !isServerTrusted && o.Cert != "" {
```

Comment:
Do we need options for automated clients that what to force or skip some of the prompts/checks in this code?

Reply:
Not sure. Automated clients can always provide their own certs, which bypasses everything. I could also add a --yes flag.

Reply:
Added --yes flag